code/game/objects/items/reagent_containers/hypospray.dm

```diff
@@ -215,69 +217,22 @@
 	init_reagent_flags = REFILLABLE|DRAINABLE
 	liquifier = TRUE
 
+/obj/item/reagent_containers/hypospray/ui_data(mob/user)
+	var/list/data = list()
 
-/obj/item/reagent_containers/hypospray/open_ui(mob/user)
-	var/dat = {"
-	<B><A href='?src=[text_ref(src)];autolabeler=1'>Activate Autolabeler</A></B><BR>
-	<B>Current Label:</B> [label]<BR>
-	<BR>
-	<B><A href='?src=[text_ref(src)];overlayer=1'>Activate Tagger</A></B><BR>
-	<B>Current Tag:</B> [description_overlay]<BR>
-	<BR>
-	<B><A href='byond://?src=[text_ref(src)];inject_mode=1'>Toggle Mode:</A></B><BR>
-	<B>Current Mode:</B> [inject_mode ? "Inject" : "Draw"]<BR>
-	<BR>
-	<B><A href='byond://?src=[text_ref(src)];set_transfer=1'>Set Transfer Amount:</A></B><BR>
-	<B>Current Transfer Amount [amount_per_transfer_from_this]</B><BR>
-	<BR>
-	<B><A href='byond://?src=[text_ref(src)];flush=1'>Empty Hypospray:</A></B><BR>
-	<BR>"}
-
-	var/datum/browser/popup = new(user, "hypospray")
-	popup.set_content(dat)
-	popup.open()
-
-
-/obj/item/reagent_containers/hypospray/advanced/open_ui(mob/user)
-	var/dat = {"
-	<B><A href='?src=[text_ref(src)];autolabeler=1'>Activate Autolabeler</A></B><BR>
-	<B>Current Label:</B> [label]<BR>
-	<BR>
-	<B><A href='?src=[text_ref(src)];overlayer=1'>Activate Tagger</A></B><BR>
-	<B>Current Tag:</B> [description_overlay]<BR>
-	<BR>
-	<B><A href='byond://?src=[text_ref(src)];inject_mode=1'>Toggle Mode:</A></B><BR>
-	<B>Current Mode:</B> [inject_mode ? "Inject" : "Draw"]<BR>
-	<BR>
-	<B><A href='byond://?src=[text_ref(src)];set_transfer=1'>Set Transfer Amount:</A></B><BR>
-	<B>Current Transfer Amount:</B> [amount_per_transfer_from_this]<BR>
-	<BR>
-	<B><A href='byond://?src=[text_ref(src)];displayreagents=1'>Display Reagent Content:</A></B><BR>
-	<BR>
-	<BR>
-	<B><A href='byond://?src=[text_ref(src)];flush=1'>Empty Hypospray:</A></B><BR>
-	<BR>"}
-
-	var/datum/browser/popup = new(user, "hypospray")
-	popup.set_content(dat)
-	popup.open()
-
-
-/obj/item/reagent_containers/hypospray/Topic(href, href_list)
+	data["InjectMode"] = inject_mode
+	data["CurrentLabel"] = label
+	data["CurrentTag"] = description_overlay
+	data["TransferAmount"] = amount_per_transfer_from_this
+
+	return data
+
+/obj/item/reagent_containers/hypospray/ui_act(action, list/params, datum/tgui/ui, datum/ui_state/state)
 	. = ..()
 	if(.)
 		return
 
-	if(href_list["inject_mode"])
-		if(inject_mode)
-			to_chat(usr, span_notice("[src] has been set to draw mode. It will now drain reagents."))
-
-		else
-			to_chat(usr, span_notice("[src] has been set to inject mode. It will now inject reagents."))
-		inject_mode = !inject_mode
-		update_icon()
-
-	else if(href_list["autolabeler"])
+	if(action == "ActivateAutolabeler")
```

Comment:
All of this action checking should be a switch, like all the other ui_acts

Reply:
oh right, I looked at other TGUI and they just had it like this, I completely forgot switches exist.